File: backend/app/desktop_api.py
from .kmeans_runtime import ConfigError, KMeansRuntime
from .window_controller import WindowController

__all__ = ["ConfigError", "DesktopApi"]


class DesktopApi:
    """pywebview js_api surface: a thin facade over KMeansRuntime + WindowController."""

    def __init__(self):
        self.runtime = KMeansRuntime()
        self.window_controller = WindowController()

    def bind_window(self, window):
        self.window_controller.bind(window)

    def window_begin_drag(self):
        self.window_controller.begin_drag()

    def window_minimize(self):
        self.window_controller.minimize()

    def window_toggle_maximize(self):
        self.window_controller.toggle_maximize()

    def window_close(self):
        self.window_controller.close()

    def ready(self):
        return self.runtime.ready()

    def get_source(self, engine_name: str):
        return self.runtime.get_source(engine_name)

    def configure(self, cfg: dict):
        return self.runtime.configure(cfg)

    def get_points_bytes(self):
        return self.runtime.get_points_bytes()

    def start(self):
        return self.runtime.start()

    def pause(self):
        return self.runtime.pause()

    def resume(self):
        return self.runtime.resume()

    def step(self):
        return self.runtime.step()

    def reset(self):
        return self.runtime.reset()

    def get_frame_bytes(self, engine_name: str):
        return self.runtime.get_frame_bytes(engine_name)

    def get_stats(self, engine_name: str):
        return self.runtime.get_stats(engine_name)

    def get_benchmark_result(self):
        return self.runtime.get_benchmark_result()

    def shutdown(self):
        self.runtime.shutdown()

File: frontend/src/lib/desktopApi.ts
import type { Config } from "./types";

export interface DesktopApi {
  ready(): Promise<{ gpu_available: boolean; gpu_info: unknown }>;
  configure(cfg: Config): Promise<Record<string, unknown>>;
  start(): Promise<void>;
  pause(): Promise<void>;
  resume(): Promise<void>;
  step(): Promise<void>;
  reset(): Promise<Record<string, unknown> | null>;
  get_stats(engine: "cpu" | "gpu"): Promise<Record<string, unknown> | null>;
  get_source(engine: "cpu" | "gpu"): Promise<{ code: string; assign_range: [number, number]; update_range: [number, number] } | null>;
  get_benchmark_result(): Promise<Record<string, unknown> | null>;
  window_begin_drag(): Promise<void>;
  window_minimize(): Promise<void>;
  window_toggle_maximize(): Promise<void>;
  window_close(): Promise<void>;
}

declare global {
  interface Window {
    pywebview?: { api: DesktopApi };
  }
}
